compiler/rustc_middle/src/ty/_match.rs

```diff
@@ -131,3 +131,72 @@ impl<'tcx> TypeRelation<'tcx> for Match<'tcx> {
         Ok(a.rebind(self.relate(a.skip_binder(), b.skip_binder())?))
     }
 }
+
+pub struct CollectAllMismatches<'tcx> {
+    pub tcx: TyCtxt<'tcx>,
+    pub param_env: ty::ParamEnv<'tcx>,
+    pub errors: Vec<TypeError<'tcx>>,
+}
+
+impl<'tcx> TypeRelation<'tcx> for CollectAllMismatches<'tcx> {
+    fn tag(&self) -> &'static str {
+        "CollectAllMismatches"
+    }
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+    fn intercrate(&self) -> bool {
+        false
+    }
+    fn param_env(&self) -> ty::ParamEnv<'tcx> {
+        self.param_env
+    }
+    fn a_is_expected(&self) -> bool {
+        true
+    } // irrelevant
+    fn mark_ambiguous(&mut self) {
+        bug!()
+    }
+    fn relate_with_variance<T: Relate<'tcx>>(
+        &mut self,
+        _: ty::Variance,
+        _: ty::VarianceDiagInfo<'tcx>,
+        a: T,
+        b: T,
+    ) -> RelateResult<'tcx, T> {
+        self.relate(a, b)
+    }
+    fn regions(
+        &mut self,
+        a: ty::Region<'tcx>,
+        _b: ty::Region<'tcx>,
+    ) -> RelateResult<'tcx, ty::Region<'tcx>> {
+        Ok(a)
+    }
+    fn tys(&mut self, a: Ty<'tcx>, b: Ty<'tcx>) -> RelateResult<'tcx, Ty<'tcx>> {
+        if a == b || matches!(a.kind(), ty::Infer(_)) || matches!(b.kind(), ty::Infer(_)) {
```

Comment:
This should probably use `InferCtxt::can_eq` instead of these hard-coded cases and `Ty` equality. We'll need to pass in an `InferCtxt`, but it should be more robust that way.

Reply:
rustc_middle doesn't depend on rustc_infer :-/

Reply:
(or even in `compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs` itself, so people aren't tempted to use it without understanding what this relation does, hehe)

Reply:
Done!

Reply:
Ah, right. The matches here I added to avoid the following super_relate_tys from blowing up and not to make them "equal". :-/

Reply:
haha, that's because we fail to relate a concrete type against an *integer* inference variable. You probably should be using `super_combine_tys` instead of `super_relate_tys` to handle this case, and putting it inside of a probe. For example: `Err(Sorts(ExpectedFound { expected: A, found: {integer} }))`

But it seems like there's a lot of back and forth with my suggestions causing ICEs, so if you don't want to fix this then I'll do it as a follow-up, since `{integer} == ConcreteType` is possibly going to lead to some weird diagnostics.